subprojects/core/src/main/java/org/gradle/internal/execution/TaskGraphBuildExecutionAction.java

```diff
@@ -39,28 +40,41 @@
  * A {@link BuildWorkExecutor} that does not execute any tasks, but prints the task graph instead.
  */
 public class TaskGraphBuildExecutionAction implements BuildWorkExecutor {
+    private final BuildWorkExecutor delegate;
     private final StyledTextOutputFactory textOutputFactory;
-
-    public TaskGraphBuildExecutionAction(StyledTextOutputFactory textOutputFactory) {
+    private final ConfigurationTimeBarrier configurationTimeBarrier;
+
+    public TaskGraphBuildExecutionAction(
+        BuildWorkExecutor delegate,
+        StyledTextOutputFactory textOutputFactory,
+        ConfigurationTimeBarrier configurationTimeBarrier
+    ) {
+        this.delegate = delegate;
         this.textOutputFactory = textOutputFactory;
+        this.configurationTimeBarrier = configurationTimeBarrier;
     }
 
     @Override
     public ExecutionResult<Void> execute(GradleInternal gradle, FinalizedExecutionPlan plan) {
+        if (configurationTimeBarrier.isAtConfigurationTime()) {
+            return delegate.execute(gradle, plan);
+        }
         StyledTextOutput output = textOutputFactory.create(TaskGraphBuildExecutionAction.class);
 
-        plan.getContents().getScheduledNodes().visitNodes((nodes, entryNodes) -> {
-            String invocation = gradle
+        if (gradle.isRootBuild()) {
```

Comment:
💭  Let's add a comment on why this check is necessary. I'd highlight that the task sub-graph from an included build will be traversed and printed from the root build as well.